### PR TITLE
[DNM] Make PG::queue_op lockless for common case

### PR DESCRIPTION
**Pull request by dlambrig**
PG::queue_op won't queue a request if the epoch is newer
than the one in the request. This patch changes that check
to be done in a lockless manner.

This is a  WIP patch, uploaded so other's could take a peek. 
more testing / analysis needed.

(1) obtain osdmap using an RCU managed pointer
(2) manage the osdmap's epoch using an atomic counter
(3) maintain request ordering for the session

Signed-off-by: Dan Lambright <dlambrig@gmail.com>

### CODE REVIEW

**Comment by dachary** [comment]
jenkins test this please (no log)

**Comment by dachary** [comment]
jenkins test this please (no log)

File: CMakeLists.txt [hunk 1]
```diff
@@ -182,16 +182,23 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ${ENABLE_SHARED})
 
 find_package(Backtrace)
 
+find_package(urcu REQUIRED)
```
Comment (mattbenjamin) [inline CMakeLists.txt]: @dlambrig looks like we have an unconditional find_package(urcu), and again in if(LINUX);  only the LINUX block sets HAVE_URCU;  is the LINUX check needed now?

File: CMakeLists.txt [hunk 2]
```diff
@@ -182,16 +182,23 @@ set(CMAKE_POSITION_INDEPENDENT_CODE ${ENABLE_SHARED})
...
+  find_package(urcu REQUIRED)
+  set(HAVE_URCU ${URCU_FOUND})
+
   find_package(aio REQUIRED)
   set(HAVE_LIBAIO ${AIO_FOUND})
 
   find_package(blkid REQUIRED)
   set(HAVE_BLKID ${BLKID_FOUND})
 else()
+  set(HAVE_URCU OFF)
```
Comment (mattbenjamin) [inline CMakeLists.txt]: @dlambrig and can the !HAVE_URCU case happen now?

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -220,6 +220,8 @@ if(HAVE_XIO)
   list(APPEND EXTRALIBS ${XIO_LIBRARY} pthread rt)
 endif(HAVE_XIO)
 
+set(RCU_LIBS ${GPERFTOOLS_TCMALLOC_LIBRARY})
```
Comment (mattbenjamin) [inline src/CMakeLists.txt]: @dlambrig looks odd--and different from form used in the target_link_libraries(...) locations

File: src/common/ceph_context.h [hunk 4]
```diff
@@ -146,6 +146,8 @@ class CephContext {
    */
   CryptoHandler *get_crypto_handler(int type);
 
+  pthread_key_t registered;
```
Comment (mattbenjamin) [inline src/common/ceph_context.h]: @dlambrig could use a more specific var name?  (nit)

File: src/mon/OSDMonitor.cc [hunk 5]
```diff
@@ -138,12 +138,12 @@ void OSDMonitor::create_initial()
 void OSDMonitor::update_from_paxos(bool *need_bootstrap)
 {
   version_t version = get_last_committed();
-  if (version == osdmap.epoch)
+  if (version == osdmap.get_epoch())
```
Comment (mattbenjamin) [inline src/mon/OSDMonitor.cc]: @dlambrig change to use epoch accessor is just stylistic? (not objecting)

File: src/msg/async/AsyncConnection.cc [hunk 6]
```diff
@@ -327,6 +327,8 @@ void AsyncConnection::process()
   bool already_dispatch_writer = false;
   std::lock_guard<std::mutex> l(lock);
   last_active = ceph::coarse_mono_clock::now();
+
+  rcu_quiescent_state();
```
Comment (mattbenjamin) [inline src/msg/async/AsyncConnection.cc]: maybe add a comment above the rcu_quiescent_state() call noting its function?

File: src/msg/async/Event.cc [hunk 7]
```diff
@@ -331,11 +332,18 @@ int EventCenter::process_events(int timeout_microseconds)
 
   ldout(cct, 10) << __func__ << " wait second " << tv.tv_sec << " usec " << tv.tv_usec << dendl;
   vector<FiredFileEvent> fired_events;
+
+  rcu_thread_offline();
```
Comment (mattbenjamin) [inline src/msg/async/Event.cc]: similarly here;  I understand this logic pretty well from NEBR (an epoch-based reclaimation variant), but find the offline() online() notation surprising and might not have been 100% sure of the behavior if I were just encountering it

File: src/msg/simple/Pipe.cc [hunk 8]
```diff
@@ -1518,17 +1519,22 @@ void Pipe::stop_and_wait()
 void Pipe::reader()
 {
   pipe_lock.Lock();
-
+
```
Comment (mattbenjamin) [inline src/msg/simple/Pipe.cc]: ws?

File: src/osd/OSDMap.cc [hunk 9]
```diff
@@ -874,7 +874,7 @@ void OSDMap::Incremental::generate_test_instances(list<Incremental*>& o)
 
 void OSDMap::set_epoch(epoch_t e)
 {
-  epoch = e;
```
Comment (mattbenjamin) [inline src/osd/OSDMap.cc]: @dlambrig epoch is an RCU var?

**Comment by dlambrig** [comment]
Have moved this PR to branch wip-lockless-queue-op.

**Pull request closed without merging**